akka-docs-dev/rst/java/http/server-side/websocket-support.rst

```diff
@@ -3,6 +3,111 @@
+i.e. a sequence of octets or a text message, i.e. a sequence of unicode code points.
+
+Akka HTTP provides a straight-forward model for this abstraction:
+
+.. includecode:: /../../akka-http-core/src/main/scala/akka/http/javadsl/model/ws/Message.scala
+   :include: message-model
+
+The data of a message is provided as a stream because WebSocket messages do not have a predefined size and could
+(in theory) be infinitely long. However, only one message can be open per direction of the WebSocket connection,
+so that many application level protocols will want to make use of the delineation into (small) messages to transport
+single application-level data units like "one event" or "one chat message".
+
+Many messages are small enough to be sent or received in one go. As an opportunity for optimization, the model provides
+a ``Strict`` subclass for each kind of message which contains data as a strict, i.e. non-streamed, ``ByteString`` or
+``String``.
+
+For sending data, the ``Strict`` variant is often the natural choice when the complete message has already been
+assembled. While receiving data from the network connection the WebSocket implementation tries to create a ``Strict``
+message whenever possible, i.e. when the complete data was received in one chunk. However, the actual chunking
+of messages over a network connection and through the various streaming abstraction layers is not deterministic from
+the perspective of the application. Therefore application code must be able to handle both streaming and strict messages
+and not expect certain messages to be strict. (Particularly, note that tests against ``localhost`` will behave
+differently from when data is received over a physical network connection.)
+
+
+Core-level support
+------------------
+
+On the server-side a request to upgrade the connection to WebSocket is provided through a special header that is added
+to a request by the implementation. Whenever a request contains the synthetic
+``akka.http.javadsl.model.ws.UpgradeToWebsocket``-header an HTTP request was a valid WebSocket upgrade request.
+Methods on this header can be used to create a response that will upgrade the connection to a WebSocket connection and
+install a ``Flow`` to handle WebSocket traffic on this connection.
+
+There's a helper method ``akka.http.javadsl.model.ws.Websocket.handleWebsocketRequestWith`` which can be used if
+only Websocket requests are expected. It looks for the ``UpgradeToWebsocket`` header and returns a response that will
+install the passed Websocket handler if the header is found. If the request is no Websocket request it will return
+a ``400 Bad Request`` error response.
+
+The following example shows how to handle a WebSocket request using just the low-level http-core API:
+
+.. includecode:: ../../code/docs/http/javadsl/server/WebsocketCoreExample.java
+   :include: websocket-example-using-core
```

Comment:
Yes, I agree it's somewhat overlong. I'll see how I can distill it so that the relevant pieces are still displayed on the documentation page.